internal/expr/parser.go

```
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+		}
+
+		idField, ok := p.parseIdentifierAsterisk()
+		if !ok {
+			return fullName{}, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
-		return fullName{}, false,
-			fmt.Errorf("go object near char %d not qualified", p.pos)
+		return fullName{id, idField}, true, nil
 	}
+
 	cp.restore()
 	return fullName{}, false, nil
 }
 
-// parseInputExpression parses an input expression of the form $Type.name.
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+	cp := p.save()
+
+	if !p.skipByte('(') {
+		return nil, false, nil
+	}
+
+	parenPos := p.pos
+	p.skipSpaces()
+
+	nextItem := true
+	var objs []fullName
+	for i := 0; nextItem; i++ {
+		if obj, ok, err := parseFn(p); ok {
+			objs = append(objs, obj)
+		} else if err != nil {
+			return nil, false, err
+		} else if i == 0 {
+			// If the first item is not what we are looking for, we exit.
+			cp.restore()
+			return nil, false, nil
+		} else {
+			// On subsequent items we return an error.
+			return nil, false, fmt.Errorf("column %d: invalid expression", p.pos)
+		}
+
+		p.skipSpaces()
+		if p.skipByte(')') {
+			return objs, true, nil
+		}
+
+		nextItem = p.skipByte(',')
+		p.skipSpaces()
+	}
+	return nil, false, fmt.Errorf("column %d: missing closing parentheses", parenPos)
+}
+
+// parseColumns parses a list of columns. For lists of more than one column the
+// columns must be enclosed in brackets e.g. "(col1, col2) AS &Person.*".
+func (p *Parser) parseColumns() ([]fullName, bool) {
+	cp := p.save()
+
+	// Case 1: A single column e.g. "p.name".
+	if col, ok, _ := p.parseColumn(); ok {
+		return []fullName{col}, true
+	}
+
+	// Case 2: Multiple columns e.g. "(p.name, p.id)".
+	if cols, ok, _ := p.parseList((*Parser).parseColumn); ok {
+		return cols, true
+	}
+
+	cp.restore()
+	return nil, false
```

Would this method need to return an error when both `parseColumn()` and `parseList()` do not return a happy result? i.e. when both Case 1 and Case 2 are not applicable. Or, similar to what's done at `parseTargets()` below, would it need to return within Case 1 and Case 2 when an error occurs?

---

See comment above.